cerbos/sdk/_async/_grpc.py

```diff
@@ -0,0 +1,658 @@
+from cerbos.response.v1 import response_pb2
+from cerbos.schema.v1 import schema_pb2
+from cerbos.sdk.grpc.utils import get_resource, is_allowed
+from cerbos.sdk.model import CerbosTLSError, CerbosTypeError
+from cerbos.svc.v1 import svc_pb2_grpc
+
+_PLAYGROUND_INSTANCE_KEY = "playground-instance"
+
+_default_paths = ssl.get_default_verify_paths()
+TLSVerify = Union[str, bool]
+
+
+# TODO(saml) type errors generated from passing incorrect types to proto generated code currently
+# aren't great, e.g. passing the incorrect Principal type to CheckResourcesRequest results in:
+#     "Message must be initialized with a dict: cerbos.request.v1.CheckResourcesRequest".
+# Investigate more useful TypeError returns.
+def handle_errors(method):
+    @wraps(method)
+    async def wrapper(*args, **kwargs):
+        try:
+            return await method(*args, **kwargs)
+        except TypeError as e:
+            raise CerbosTypeError(str(e))
+
+    return wrapper
+
+
+def get_cert(c: TLSVerify) -> bytes | None:
+    try:
+        if isinstance(c, str):
+            with open(c, "rb") as f:
+                return f.read()
+        elif isinstance(c, bool):
+            # Attempt to retrieve the cert from the location specified at `SSL_CERT_FILE`
+            # or the default location if not specified.
+            filename = _default_paths.cafile
+            if cf := os.getenv("SSL_CERT_FILE"):
+                filename = cf
+            with open(filename, "rb") as f:
+                return f.read()
+    except IOError:
+        raise CerbosTLSError(f"Error reading certificate from file: {c}")
+    except Exception:
+        raise CerbosTLSError("Error retrieving certificate")
+
+    raise TypeError("TLSVerify should be a string or boolean")
+
+
+class PlaygroundInstanceCredentials(grpc.AuthMetadataPlugin):
+    def __init__(self, playground_instance: str):
+        self._playground_instance = playground_instance
+
+    def __call__(
+        self,
+        context: grpc.AuthMetadataContext,
+        callback: grpc.AuthMetadataPluginCallback,
+    ) -> None:
+        callback(((_PLAYGROUND_INSTANCE_KEY, self._playground_instance),), None)
+
+
+class AsyncClientBase:
+    _channel: grpc.aio.Channel
+
+    def __init__(
+        self,
+        host: str,
+        creds: grpc.ChannelCredentials,
+        methods: List[Dict[str, str]] = None,
+        tls_verify: TLSVerify = False,
+        timeout_secs: float | None = None,
+        request_retries: int = 0,
+        wait_for_ready: bool = False,
+    ):
+        if timeout_secs and not isinstance(timeout_secs, (int, float)):
+            raise TypeError("timeout_secs must be a number type")
+
+        if request_retries and not isinstance(request_retries, (int, float)):
+            raise TypeError(
+                "request_retries must be a number type. anything below 2 is treated as 0 (disabled)"
+            )
+
+        if request_retries < 2:
+            request_retries = 0
```

Comment:
Is it not possible to retry once?

Reply:
Interestingly, it counts the initial attempt as one of the retries, [so passing 1 is invalid](https://github.com/grpc/grpc-proto/blob/ec886024c2f7b7f597ba89d5b7d60c3f94627b17/grpc/service_config/service_config.proto#L133).